### DISCUSSION TITLE
CompiledBinding from Code

### DISCUSSION THREAD

**Discussion by danielmayost** [AvaloniaUI/Avalonia#10077] [Q&A] [answered]
Does anybody know how to create compiled binding (like from xaml) from code?

**Comment by amwx** [comment, marked as answer]
It's *really* ugly, but this seems to work (example binding to the Title property on a Window):

Test view model class
```C# 
public class TestClass
{
    public string Name { get; set; } = "Test CompiledBinding From Code";
}
```

Construct the binding
```C#
var x = new CompiledBindingPathBuilder();
x = x.SetRawSource(new TestClass());
x = x.Property(
    new ClrPropertyInfo(
        nameof(TestClass.Name), 
        obj => ((TestClass)obj).Name,
        (obj, val) => ((TestClass)obj).Name = (string)val, 
        typeof(string)),
        (weakRef, iPropInfo) =>
        {
            return PropertyInfoAccessorFactory.CreateInpcPropertyAccessor(weakRef, iPropInfo);
        });
var path = x.Build();

var cb = new CompiledBindingExtension(path);
var ib = cb.Initiate(this, TitleProperty);
BindingOperations.Apply(this, TitleProperty, ib, null);
```

![image](https://user-images.githubusercontent.com/40413319/215372597-285a621e-6244-4ed7-98c1-6b05d493328f.png)

**Reply by danielmayost** [reply]
Wow, thank. Actually I need it to create a `MarkupExtension`, And only the definition of C`ompiledBindingPathBuilder` was missing for me.

**Reply by uno-studio** [reply]
> Nó _thực sự_ xấu, nhưng điều này có vẻ hiệu quả (ví dụ liên kết với thuộc tính Tiêu đề trên Cửa sổ):
> 
> Kiểm tra xem lớp mô hình
> 
> ```cs
> public class TestClass
> {
>     public string Name { get; set; } = "Test CompiledBinding From Code";
> }
> ```
> 
> Xây dựng ràng buộc
> 
> ```cs
> var x = new CompiledBindingPathBuilder();
> x = x.SetRawSource(new TestClass());
> x = x.Property(
>     new ClrPropertyInfo(
>         nameof(TestClass.Name), 
>         obj => ((TestClass)obj).Name,
>         (obj, val) => ((TestClass)obj).Name = (string)val, 
>         typeof(string)),
>         (weakRef, iPropInfo) =>
>         {
>             return PropertyInfoAccessorFactory.CreateInpcPropertyAccessor(weakRef, iPropInfo);
>         });
> var path = x.Build();
> 
> var cb = new CompiledBindingExtension(path);
> var ib = cb.Initiate(this, TitleProperty);
> BindingOperations.Apply(this, TitleProperty, ib, null);
> ```
> 
> ![hình ảnh](https://user-images.githubusercontent.com/40413319/215372597-285a621e-6244-4ed7-98c1-6b05d493328f.png)

Can you help me convert this code to use CompiledBinding?
```c#
public class LocalizeExtension : MarkupExtension
{
    public LocalizeExtension(string key)
    {
        Key = key;
    }

    public string Key { get; set; }

    public string Context { get; set; }

    public override object ProvideValue(IServiceProvider serviceProvider)
    {
        var keyToUse = Key;
        if (!string.IsNullOrWhiteSpace(Context))
            keyToUse = $"{Context}/{Key}";
        var binding = new ReflectionBindingExtension($"[{keyToUse}]")
        {
            Mode = BindingMode.OneWay,
            Source = Localizer.Instance
        };
        return binding.ProvideValue(serviceProvider);
    }
}

```
